### ISSUE TITLE
FP in `collapsible_match` on futures `select!` macro

### ISSUE THREAD

**Issue by nipunn1313** [rust-lang/rust-clippy#7591]
I tried this code:

```rust
use futures::executor::block_on;
use futures::future;
use futures::select;

fn main() {
    let mut a = future::ready(Some(4));
    let mut b = future::pending::<()>();

    block_on(async {
        let mut res = 0;
        select! {
            a_res = a => if let Some(a_res) = a_res { res = a_res + 1 },
            _ = b => (),
        };
        assert_eq!(res, 5);
    })
}
```

I expected to see no clippy error

Instead, this happened: 
```
warning: unnecessary nested `if let` or `match`
  --> src/main.rs:12:26
   |
12 |             a_res = a => if let Some(a_res) = a_res { res = a_res + 1 },
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::collapsible_match)]` on by default
help: the outer pattern can be modified to include the inner pattern
  --> src/main.rs:12:13
   |
12 |             a_res = a => if let Some(a_res) = a_res { res = a_res + 1 },
   |             ^^^^^               ^^^^^^^^^^^ with this pattern
   |             |
   |             replace this binding
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_match

```

clippy seems to want a different code which is not valid:
```rust
 Some(a_res) = a => res = a_res + 1,
```
leading to
```
  --> src/main.rs:11:9
   |
11 | /         select! {
12 | |             Some(a_res) = a => res = a_res + 1,
13 | |             _ = b => (),
14 | |         };
   | |          ^
   | |          |
   | |          pattern `_0(None)` not covered
   | |__________`main::{closure#0}::__PrivResult<std::option::Option<i32>, ()>` defined here
   |            not covered
```

### Meta

```
➜  clippy_test cargo clippy -V
clippy 0.1.56 (0035d9dc 2021-08-16)
```

### It's a recent regression

Notably, I found that with
```
clippy 0.1.56 (2d2bc94c 2021-08-15)
```

things work fine

came up in the testsuite of https://github.com/rust-lang/futures-rs/pull/2479

**Comment by xFrednet** [comment]
Thank you for the report. At a first glance, it looks like the lint doesn't check if the code is inside a macro (I haven't looked at the linting code, though). :upside_down_face: 

@rustbot label +I-false-positive

**Comment by nipunn1313** [comment]
#7565 looks potentially suspicious as it contains an August 16th change to the collapsible_match lint

https://github.com/rust-lang/rust-clippy/commits/master/clippy_lints/src/collapsible_match.rs

The timing lines up.

**Comment by xFrednet** [comment]
There have also been some changes in the Rust repo on this specific lint. I wouldn't think that it originated from #7565 as your Clippy version stated a date before that PR was merged. :upside_down_face:

**Comment by nipunn1313** [comment]
Yep - confirmed @xFrednet's hypothesis - by running `master` of clippy with various older/newer toolchains. It appears to be a change in the rust repo.

**Comment by xFrednet** [comment]
Thank you, @nipunn1313, for the test :upside_down_face:.

cc. @camsteffen in case you want to work on another regression issue :wink:
